kimchi/src/public_input_only_prover.rs

```diff
@@ -0,0 +1,627 @@
+    proof::{
+        PointEvaluations, ProofEvaluations, ProverCommitments, ProverProof, RecursionChallenge,
+    },
+    verifier_index::VerifierIndex,
+};
+use ark_ec::ProjectiveCurve;
+use ark_ff::{Field, One, PrimeField, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Polynomial,
+    Radix2EvaluationDomain as D, UVPolynomial,
+};
+use mina_poseidon::{sponge::ScalarChallenge, FqSponge};
+use o1_utils::ExtendedDensePolynomial as _;
+use once_cell::sync::OnceCell;
+use poly_commitment::{
+    commitment::{absorb_commitment, b_poly_coefficients, BlindedCommitment, PolyComm},
+    evaluation_proof::DensePolynomialOrEvaluations,
+    srs::{endos, SRS},
+};
+use rand::{CryptoRng, RngCore};
+use std::array;
+use std::sync::Arc;
+
+/// The result of a proof creation or verification.
+type ProverResult<T> = std::result::Result<T, ProverError>;
+
+/// Creates a hard-coded verifier index that proofs generated by
+/// `create_recursive_public_input_only` will satisfy.
+pub fn verifier_index<G: KimchiCurve>(
+    srs: Arc<SRS<G>>,
+    domain: EvaluationDomains<G::ScalarField>,
+    num_public_inputs: usize,
+    num_prev_challenges: usize,
+) -> VerifierIndex<G> {
+    let shifts = permutation::Shifts::new(&domain.d1);
+    let (endo_q, _endo_r) = endos::<G::OtherCurve>();
+
+    // FeatureFlags with all features disabled.
+    let feature_flags: FeatureFlags = std::default::Default::default();
+
+    let (linearization, powers_of_alpha) =
+        crate::linearization::expr_linearization(Some(&feature_flags), true);
+
+    let make_comm = |comm| PolyComm {
+        unshifted: vec![comm],
+        shifted: None,
+    };
+
+    VerifierIndex {
+        domain: domain.d1,
+        max_poly_size: srs.g.len(),
+        srs: srs.clone().into(),
+        public: num_public_inputs,
+        prev_challenges: num_prev_challenges,
+
+        sigma_comm: array::from_fn(|i| PolyComm {
+            // Encodes the polynomial `f(x) = x * shifts[i]`, with no blinding.
+            // This represents the identity permutation.
+            unshifted: vec![srs.g[1].mul(shifts.shifts[i]).into_affine()],
+            shifted: None,
+        }),
+        coefficients_comm: array::from_fn(|i| {
+            make_comm(if i == 0 {
+                // The polynomial `f(x) = 1`, without blinding.
+                srs.g[0]
+            } else {
+                // The polynomial `f(x) = 0`, with blinding factor 1.
+                // This blinding allows us to represent the commitment in affine coordinates.
+                srs.h
+            })
+        }),
+        // The polynomial `f(x) = 1`, without blinding.
+        // The generic gate is enabled on every row; combined with the coefficients, and with the
+        // mixin for public inputs, this encodes the equation `witness[0] = public_input`.
+        // Generic gate's equation is:
+        //    c_0 \cdot l + c_1 \cdot r + c_2 \cdot o + c_3 \cdot (l \times r) + c_4
```

Comment:
@mrmr1993 could you please double check this comment I added is what you had in mind?